addons/lunch/models/lunch_order.py

```diff
@@ -41,10 +41,128 @@ class LunchOrder(models.Model):
+            line.available_toppings_2 = bool(line.env['lunch.topping'].search_count([('category_id', '=', line.category_id.id), ('topping_category', '=', 2)]))
+            line.available_toppings_3 = bool(line.env['lunch.topping'].search_count([('category_id', '=', line.category_id.id), ('topping_category', '=', 3)]))
+
     def init(self):
         self._cr.execute("""CREATE INDEX IF NOT EXISTS lunch_order_user_product_date ON %s (user_id, product_id, date)"""
             % self._table)
 
+
+    def _extract_toppings(self, values):
+        """
+            If called in api.multi then it will pop topping_ids_1,2,3 from values
+        """
+        if self.ids:
+            # TODO This is not taking into account all the toppings for each individual order, this is usually not a problem
+            # since in the interface you usually don't update more than one order at a time but this is a bug nonetheless
+            topping_1 = values.pop('topping_ids_1')[0][2] if 'topping_ids_1' in values else self[:1].topping_ids_1.ids
+            topping_2 = values.pop('topping_ids_2')[0][2] if 'topping_ids_2' in values else self[:1].topping_ids_2.ids
+            topping_3 = values.pop('topping_ids_3')[0][2] if 'topping_ids_3' in values else self[:1].topping_ids_3.ids
+        else:
+            topping_1 = values['topping_ids_1'][0][2] if 'topping_ids_1' in values else []
+            topping_2 = values['topping_ids_2'][0][2] if 'topping_ids_2' in values else []
+            topping_3 = values['topping_ids_3'][0][2] if 'topping_ids_3' in values else []
+
+        return topping_1 + topping_2 + topping_3
+
+    @api.constrains('topping_ids_1', 'topping_ids_2', 'topping_ids_3')
+    def _check_topping_quantity(self):
+        errors = {
+            '1_more': _('You should order at least one %s'),
+            '1': _('You have to order one and only one %s'),
+        }
+        for line in self:
+            for index in range(1, 4):
+                availability = line['available_toppings_%s' % index]
+                quantity = line['topping_quantity_%s' % index]
+                toppings = line['topping_ids_%s' % index].filtered(lambda x: x.topping_category == index)
+                label = line['topping_label_%s' % index]
+
+                if availability and quantity != '0_more':
+                    check = bool(len(toppings) == 1 if quantity == '1' else toppings)
+                    if not check:
+                        raise ValidationError(errors[quantity] % label)
+
+    @api.model
+    def create(self, values):
+        lines = self._find_matching_lines({
+            **values,
+            'toppings': self._extract_toppings(values),
+        })
+        if lines:
+            # FIXME This will update multiple lines in the case there are multiple matching lines
+            # which should not happen trough the interface
+            lines.update_quantity(1)
+            return lines[:1]
```

Comment:
This return is weird, but create is expected to return a recordset which contains only one id